### ISSUE TITLE
Can a function return jslt expression or another function

### ISSUE THREAD

**Issue by darshannahar** [schibsted/jslt#165]
Is JSLT expression considered as a first-class jslt type? Can I have an extension function return a jslt expression or can I have a jslt function return another jslt function?

**Comment by larsga** [comment]
> Is JSLT expression considered as a first-class jslt type?

No. The only things that are considered acceptable values in JSLT are JSON values.

> Can I have an extension function return a jslt expression or can I have a jslt function return another jslt function?

Sorry, no.

**Comment by darshannahar** [comment]
Thanks Lars! What would it take to make this happen? I just saw #120 requesting similar.

We have import statements which assigns variables with function. Hence, I feel it might not be completely not out of the blue.
Also, do you see any downsides to the language in doing so?

**Comment by larsga** [comment]
The biggest issue is the conceptual one, of expanding the set of values to suddenly include functions. What happens if you try to `+` functions, or use `.foo` on them, or ...? Also, how do they serialize to JSON if they get included in the output?

The second issue is the question of what the benefit is. It would be very helpful to see a convincing example of higher order functions making life easier.

**Comment by darshannahar** [comment]
The usecase for which I asked:
- Factory pattern: We serve multiple variations of our client and main jslt needs to pick util translating expression based on some context. Lets say some one needs different translator based on car model, they can do:
getMeTranslator(.model)(.data)

I could see many users leveraging this to extend their any possible translation needs. E.g. One of the other evaluation done where jslt was ruled out was ability to fetch more data asynchronously. They had a usecase where based on input data and some conditions they might want to make a IO for fetch more data. They could have implemented something like futures or promise easily with this capability combined with extension functions. SQL++ is trying hard to fit for usecases to translate json documents from disparate sources into an output view. I feel jslt could solve some of those usecses where users don't want to rely on sql++ planners or find that type of a language too limiting.

The issues you raise are valid especially the serialization one. We would definitely need helper methods like isJSLTFunction or isJSON for the users. Some suggestions:
- JSLT functions should serialize into file containing jslt expression (if someone attempts) and json should serialize into a json.
- Operators will throw errors when not used appropriately. Today if I use + or .foo on a function name I get an error which might need to continue.

**Comment by darshannahar** [comment]
@larsga Please let us know your thoughts? Do you want to permit this enhancement or we feel it is not the right enhancement?

**Comment by larsga** [comment]
I'm still on the fence about this, to be honest. It's possible to solve the technical issues, but I'm not sure it's worth the complexity.

The use case you describe looks like it could easily be handled by doing:

```
def pickTranslator(data)
   if (somecondition(...))
     function1(data)
   else if (someothercondition(...))
     function2(data)
   // ....
```

If you really need to separate the declaration of the functions from the picker you could use separate modules. We have similar use cases in my last two places of work and there we instead build the processing pipeline logic in YAML config files which either embed or refer to JSLT for filters and transforms. For our purposes that has worked far better than trying to do everything in JSLT.

It would be interesting to hear other opinions on this, too.
